### PR TITLE
[layer] Move distribute to LayerNode

### PR DESCRIPTION
**Pull request by kparichay**
move distribute to LayerNode from layer
Further, the distribute layer is now supposed to be used
as a wrapper of a layer maintained by the LayerNode,
and must not be maintained externally.
Rather, use getDistribute() to check if the node is distributed
and treat the layer as a regular layer.

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Comment by kparichay** [comment]
wait for https://github.com/nnstreamer/nnstreamer/pull/3326 for the CI to pass.

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/graph/network_graph.cpp [hunk 1]
```diff
@@ -121,8 +114,14 @@ void NetworkGraph::countNonTrainableLayersAtBegin() {
   skip_non_trainable_layers = graph.size();
 }
 
-int NetworkGraph::realizeMultiInputType(Layer &current) {
+int NetworkGraph::realizeMultiInputType(
+  const std::shared_ptr<LayerNode> &in_node) {
+  Layer &current = *in_node->getObject();
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:119]: should this be const Layer?

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 2]
```diff
@@ -139,7 +140,6 @@ class nntrainer_abstractLayer : public ::testing::Test {
                    const nntrainer::TensorDim &dim) {
     nntrainer::Tensor golden(dim);
     loadFile(expected, golden);
-    /** FIXME: golden.length() is possibly 0 many times, verify and fix this */
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_layers.cpp:142]: Is this resolved?

**Review by zhoonit — APPROVED** [review]
LGTM with minor questions

File: nntrainer/graph/network_graph.cpp [hunk 3]
```diff
@@ -203,19 +206,16 @@ int NetworkGraph::realizeActivationType(Layer &current) {
...
-  layer->setActivation(act);
   graph.ensureName(*lnode, current.getName());
 
-  if (current.getType() == TimeDistLayer::type) {
-    std::string unit_str = layer->getName();
-    graph.ensureName(*lnode, "", "_unit");
-    layer = distributeLayer(layer);
-    lnode = std::make_shared<LayerNode>(layer);
-    layer->setName(unit_str);
+  if (in_node->getDistribute()) {
```
Comment (jijoongmoon) [inline nntrainer/graph/network_graph.cpp:211]: Could you check if it is ok to remove setActivation?

**Comment by jijoongmoon** [comment]
I'm not sure if this works as we expected. Let's check again.

File: test/unittest/unittest_nntrainer_layers.cpp (hunk 2 above)
Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:142]: this was resolved with #1181

File: nntrainer/graph/network_graph.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:211]: yes, thats the expection. `distribute=true` is supposed to wrap the current layer with distribute, and any other layer it creates like activation. 

> I guess. lnode is activation layer. we need another distributed layer which will contain these activation layer.

`lnode->setProperty({"distribute=true"})` does exactly this. It creates a layer internally, and sets it in the distribute layer.

[the remark this answers is not in the document]
Comment (jijoongmoon) [reply nntrainer/graph/network_graph.cpp:211]: Understood from off-line discussion. Thanks.

**Pull request merged**